src/lib.rs

```diff
@@ -1958,25 +1958,19 @@ impl Build {
                     let mut parts = target.split('-');
                     if let Some(arch) = parts.next() {
                         let arch = &arch[5..];
-                        if target.contains("linux") && arch.starts_with("64") {
-                            cmd.args.push(("-march=rv64gc").into());
-                            cmd.args.push("-mabi=lp64d".into());
-                        } else if target.contains("freebsd") && arch.starts_with("64") {
-                            cmd.args.push(("-march=rv64gc").into());
-                            cmd.args.push("-mabi=lp64d".into());
-                        } else if target.contains("openbsd") && arch.starts_with("64") {
-                            cmd.args.push(("-march=rv64gc").into());
-                            cmd.args.push("-mabi=lp64d".into());
-                        } else if target.contains("linux") && arch.starts_with("32") {
-                            cmd.args.push(("-march=rv32gc").into());
-                            cmd.args.push("-mabi=ilp32d".into());
-                        } else if arch.starts_with("64") {
-                            cmd.args.push(("-march=rv".to_owned() + arch).into());
-                            cmd.args.push("-mabi=lp64".into());
+
+                        cmd.args.push(("-march=rv".to_owned() + arch).into());
+
+                        // If riscv32gc-* or riscv64gc-* then double-float ABI,
+                        // otherwise use soft-float ABI.
+                        let float_abi = if arch.contains("g") { "d" } else { "" };
```

Comment:
This looks fragile to me, maybe we could use `arch.ends_with("gc")` or `contains("gc")` at least.

Reply:
My reasoning is that `g` means `imafd_zifencei_zicsr`, so it implies `d` for `double`, which is required for double float ABI. Currently it seems (see table in top post) that rustc does use `lp64d`/`ilp32d` for all targets with `d`ouble. Maybe I can change the code to detect `d` or `g` and/or change the comment to reflect this reasoning better?

I guess I'll also change it so that it handles e.g. `riscv64imac_zsomething` with a multi-letter extension tha happens to contain `g`. Not sure if something other than an ISA/`-march` string appears here though tbh.

Reply:
In the RISC-V Unprivileged Specification RISC-V ISA can be decided from ISA string, which indicates that if it has D inside, the target must support double length floating points; and G means that it has IMAFD and Zifencei (F indicates Zicsr). Using C here would not cover this case correctly, as that C doesn't imply G or D.

It is possible to have an RVD implementation without RVC, for example disabling C instruction set on RV64GC platform (which is sometimes common on embedded SoC or developing new SoCs); detecting GC for D will not be correct in those cases.

Reply:
I've added a bunch of comments explaining the logic, and also updated the detection logic to handle the possible future existence of multi-letter extensions, so it won't get confused by e.g. `riscv64imac_zsomething` containing a `g`.